core-api/src/main/java/com/optimizely/ab/config/audience/match/SemanticVersion.java

```diff
@@ -0,0 +1,133 @@
+/**
+ *
+ *    Copyright 2020, Optimizely and contributors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+package com.optimizely.ab.config.audience.match;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static com.optimizely.ab.internal.AttributesUtil.parseNumeric;
+import static com.optimizely.ab.internal.AttributesUtil.stringIsNullOrEmpty;
+
+public final class SemanticVersion {
+
+    private static final String BUILD_SEPERATOR = "+";
+    private static final String PRE_RELEASE_SEPERATOR = "-";
+
+    private final String version;
+
+    public SemanticVersion(String version) {
+        this.version = version;
+    }
+
+    public int compare(SemanticVersion targetedVersion) throws Exception {
+
+        if (targetedVersion == null || stringIsNullOrEmpty(targetedVersion.version)) {
+            return 0;
+        }
+
+        String[] targetedVersionParts = splitSemanticVersion(targetedVersion.version);
+        String[] userVersionParts = splitSemanticVersion(version);
+
+        for (int index = 0; index < targetedVersionParts.length; index++) {
+
+            if (userVersionParts.length <= index) {
+                return isPreRelease(targetedVersion.version) ? 1 : -1;
+            }
+            Integer targetVersionPartInt = parseNumeric(targetedVersionParts[index]);
+            Integer userVersionPartInt = parseNumeric(userVersionParts[index]);
+
+            if (userVersionPartInt == null) {
+                // Compare strings
+                int result = userVersionParts[index].compareTo(targetedVersionParts[index]);
+                if (result != 0) {
+                    return result;
+                }
+            } else if (targetVersionPartInt != null) {
+                if (!userVersionPartInt.equals(targetVersionPartInt)) {
+                    return userVersionPartInt < targetVersionPartInt ? -1 : 1;
+                }
+            } else {
+                return -1;
+            }
+        }
+
+        if (!isPreRelease(targetedVersion.version) &&
+            isPreRelease(version)) {
+            return -1;
+        }
+
+        return 0;
+    }
+
+    private boolean isPreRelease(String semanticVersion) {
+        return semanticVersion.contains(PRE_RELEASE_SEPERATOR);
```

Comment:
remove string in isPreRelease and use as semver.isPreRelease.  the scope can be public or protected.  thanks!